Dockerfile

```diff
@@ -42,20 +54,36 @@ ENV CXX g++
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
 
-COPY . /build
+# to avoid early cache invalidation, we build only dependencies first. For this we create fresh crates we are going to overwrite.
+RUN USER=root cargo init --bin --name=mashnet-node
+RUN USER=root cargo new --lib --name=mashnet-node-runtime runtime
+# overwrite cargo.toml with real files
+COPY Cargo.toml Cargo.lock build.rs ./
+COPY ./runtime/Cargo.toml ./runtime/Cargo.lock ./runtime/
+
+# build depedencies (and bogus source files)
+RUN cargo build --release
 
-RUN /bin/bash scripts/build.sh
+# remove bogus build (but keep depedencies)
+RUN cargo clean --release -p mashnet-node-runtime
 
+# copy everything over (cache invalidation will happen here)
+COPY . /build
+# get wasm built in previous step
+COPY --from=wasm_builder /runtime/wasm/target/wasm32-unknown-unknown/release ./runtime/wasm/target/wasm32-unknown-unknown/release
+# build source again, dependencies are already built
 RUN cargo build --release
+
+# test
 RUN cargo test --release -p mashnet-node-runtime
 
 
-FROM ubuntu:xenial
+FROM debian:stretch
```

Comment:
Why not use `buster` which is the current release?

Reply:
the rust nightly image is build on stretch, wanted to keep things as similar as possible